core/src/main/java/org/elasticsearch/common/util/concurrent/ThreadContext.java

```diff
@@ -407,11 +407,10 @@ private ThreadContextStruct putHeaders(Map<String, String> headers) {
             if (headers.isEmpty()) {
                 return this;
             } else {
-                final Map<String, String> newHeaders = new HashMap<>();
+                final Map<String, String> newHeaders = new HashMap<>(this.requestHeaders);
```

Comment:
Since the method takes in `headers`, I assume those should "win" over what was in the context headers map before.  However, I did notice that in `stashAndMergeHeaders`, the `requestHeaders` are merged into the headers being passed into `putHeaders`, and the order is again such that `requestHeaders` wins over `headers`.  @s1monw can you clarify what the original intention was here?